prepreduce/prepreduce.py

```diff
@@ -0,0 +1,395 @@
+    )
+
+    parser.add_argument(
+        "-f",
+        "--force-rm-existing-target",
+        help="If the target (foo.o) already exists, remove it -- without prompting",
+        action="store_true",
+        default=False,
+    )
+
+    parser.add_argument(
+        "-F",
+        "--force-rm-existing-script",
+        help="If test.sh already exists, remove it -- without prompting",
+        action="store_true",
+        default=False,
+    )
+
+    parser.add_argument(
+        "-d",
+        "--debug",
+        help="Enable prepreduce Debug logs",
+        action="store_true",
+        default=False,
+    )
+
+    parser.add_argument(
+        "-o",
+        "--output",
+        help="Output directory of prepreduce outputs (.i files and test scripts)",
+        default=os.getcwd(),
+        type=lambda arg: validate_path_from_cli(arg, "-o/--output", True, True),
+    )
+
+    args = parser.parse_args()
+    DEBUG = args.debug
+    if DEBUG:
+        ic.enable()
+
+    validate_build_cmd_from_cli(args.build_command)
+
+    return args
+
+
+def validate_build_cmd_from_cli(build_cmd: list[str]):
+    """
+    We need V=1 to find original clang invocation
+    """
+    try:
+        _ = build_cmd.index("V=1")  # raises ValueError if not found
+    except ValueError:
+        error("Need `V=1` in your make build command (suggested to use default)")
+
+
+def validate_target_from_cli(_target: str) -> Path:
+    """
+    Allow argparse to actually validate the target argument as well as convert
+    to PosixPath
+    """
+    try:
+        posix_path = Path(_target.strip())
+    except:
+        raise argparse.ArgumentTypeError(
+            f"Invalid target provided: {_target}. Can't convert {_target} to a PosixPath. Check for typos and random symbols"
+        )
+    if posix_path.suffix != ".o" and len(posix_path.suffix):
+        error(
+            f"target file has {posix_path.suffix} extension instead of `.o`. "
+            "Either put no suffix or add `.o`. quitting."
+        )
+
+    return posix_path.with_suffix(".o")
+
+
+def validate_path_from_cli(
+    _path: str, argument: str, allow_mkdir: bool = False, do_chdir=False
+) -> Path:
+    """
+    Allow argparse to actually validate -p/--path-to-linux argument and
+    provide helpful feedback
+    """
+    try:
+        posix_path = Path(_path.strip())
+    except:
+        raise argparse.ArgumentTypeError(
+            f"Invalid {argument} argument provided: {_path} doesn't parse into a PosixPath"
+        )
+
+    if not os.path.isdir(posix_path):
+        if not allow_mkdir:
+            raise argparse.ArgumentTypeError(
+                f"Invalid {argument} argument provided: {_path} is not a directory or it doesn't exist"
+            )
+        info(f"Making directory for prepreduce's output files: {posix_path}")
+        os.mkdir(posix_path)
+    if do_chdir:
+        info(f"Changing CWD to {posix_path}")
+        os.chdir(posix_path)
```

Comment:
As far as I can tell, this is dead code.

Reply:
`do_chdir` flag is enabled here 

```py
    parser.add_argument(
        "-o",
        "--output",
        help="Output directory of prepreduce outputs (.i files and test scripts)",
        default=os.getcwd(),
        type=lambda arg: validate_path_from_cli(
            arg, "-o/--output", allow_mkdir=True, do_chdir=True
        ),
    )
```

note: just added named kwargs here. It was unreadable before with just `(..., True, True)`

Reply:
Ah, that is pretty subtle... I thought I audited all uses of `validate_path_from_cli()` but missed that. I think that it would be better to avoid `chdir` altogether and just work with absolute paths whenever possible.

Reply:
That was my initial thought as well. Had some weird behavior from `subprocess.run()` not liking my absolute path when invoking `make` (with the relative paths from the CC invocation). This may fix itself when I migrate the code over to using `kbuild`.

wip

Reply:
removed now as `chdir` seemed hacky 35639e060c54817ba32cc388d97b7725eb91b171